[Serializer] Add a `force_collection` option to `XmlEncoder`

AlexandreGerault opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 8.2
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        | 
| License       | MIT

When `XmlEncoder` decodes a tag holding a single child, it returns that child rather than a list. The shape of the decoded data therefore depends on how many children the document happens to carry, which is awkward when the XML is a source of truth and a collection may legitimately hold one entry.

```xml
<order_rows>
    <order_row><id>1</id></order_row>
</order_rows>
```

decodes to `['order_row' => ['id' => 1]]`, while the same document with two rows decodes to `['order_row' => [['id' => 1], ['id' => 2]]]`.

The new `force_collection` context option takes the list of tag names that must always decode as a collection:

```php
$decoded = $encoder->decode($xml, 'xml', [XmlEncoder::FORCE_COLLECTION => ['order_rows']]);
// ['order_row' => [['id' => 1]]] whatever the number of rows
```

It is also available on the context builder:

```php
$context = (new XmlEncoderContextBuilder())->withForceCollection(['order_rows'])->toArray();
```

The option defaults to an empty list, so documents decode exactly as before unless a tag is named.

Documentation should cover: the option and its builder method, that it is keyed by the *parent* tag name, and that it only affects decoding.

Commit:
feat(serializer): add option to force collection when decoding xml

Commit:
fix: ci and style

Commit:
ci: fix suggestions by fabbot.io

nicolas-grekas commented on src/Symfony/Component/Serializer/Encoder/XmlEncoder.php:
```diff
@@ -407,7 +422,7 @@ private function buildXml(\DOMNode $parentNode, mixed $data, string $format, arr
         }
 
-        throw new NotEncodableValueException('An unexpected value could not be serialized: '.(!\is_resource($data) ? var_export($data, true) : \sprintf('%s resource', get_resource_type($data))));
+        throw new NotEncodableValueException('An unexpected value could not be serialized: '.(!\is_resource($data) ? var_export($data, true) : \sprintf('"%s" resource', get_resource_type($data))));
```
please revert this, that's a false-positive

nicolas-grekas commented on src/Symfony/Component/Serializer/Encoder/XmlEncoder.php:
```diff
@@ -224,10 +229,20 @@ final protected function isElementNameValid(string $name): bool
+        if (\is_array($value)
+            && ($childNodeName = $node->firstChild?->nodeName)
+            && 1 === \count($value)
+            && \in_array($nodeName, $context[self::FORCE_COLLECTION] ?? [], strict: true)
```
```suggestion
            && \in_array($nodeName, $context[self::FORCE_COLLECTION] ?? [], true)
```

Commit:
revert: remove false positive from fabbot.io

Commit:
refactor(serializer): remove named parameter

Commit:
feat(serializer): use default context when context does not supply option

nicolas-grekas commented:
> we have a workaround with a custom normalizer looking at specific keys

So you figured out a way that doesn't require changing the component? then it might be the best way, because as you spotter, defining which tags in the xml tree would need this arrayification is tricky. A normalizer looks like a better place to me.

AlexandreGerault commented:
We found a workaround, but that's not exactly what I would expect from the component.

Here's the workaround we ended with, just as information:
```php
public function denormalize(mixed $data, string $type, ?string $format = null, array $context = []): mixed
{
    // If the first element is not an array, wrap it in an array
    if (! is_array(reset($data['plans']))) {
        $data['plans'] = [$data['plans']];
    }

    // The return thing
}
```

It works in our specific use case (which is why we didn't make a generic normalizer with context option and all), but I find it odd that there are no ways of telling "this XML tag should always be treated as a collection". I mean, it makes more sense to me to be handled by a decoder option, as it is (at least it's what I understand for now) directly related to the XML format. I might be wrong still.

However if I'm right (that it should be handled by the decoder), I can also understand that it would be too tricky to be handled in the decoder and that the maintainers refuses to maintain such a feature.

On the other hand, if you consider it can be added, I'd gladly work on a better implementation and look for more edge cases etc.

thomasLecler commented:
I was also looking for something like this, great 👍 
Hope this will be implemented in the coming Laravel updates

AlexandreGerault commented:
> I was also looking for something like this, great 👍 Hope this will be implemented in the coming Laravel updates

Not related to Laravel. However you can still use it as a dependency as the component is just like any other library.

thomasLecler commented:
> > I was also looking for something like this, great 👍 Hope this will be implemented in the coming Laravel updates
> 
> Not related to Laravel. However you can still use it as a dependency as the component is just like any other library

I know,

Indeed this is my case 👍

mtarld commented on src/Symfony/Component/Serializer/Encoder/XmlEncoder.php:
```diff
@@ -31,6 +31,11 @@ class XmlEncoder implements EncoderInterface, DecoderInterface, NormalizationAwa
+    /**
+     * An array of XML tags who should always be treated as a collection, even when it has only one child.
+     */
+    public const FORCE_COLLECTION = 'force_collection';
```
Could you update the `XmlEncoderContextBuilder` accordingly?

AlexandreGerault replied:
Done

mtarld commented on src/Symfony/Component/Serializer/Encoder/XmlEncoder.php:
```diff
@@ -224,10 +230,20 @@ final protected function isElementNameValid(string $name): bool
 
         $value = $this->parseXmlValue($node, $context);
 
+        if (\is_array($value)
+            && ($childNodeName = $node->firstChild?->nodeName)
+            && 1 === \count($value)
+            && \in_array($nodeName, $context[self::FORCE_COLLECTION] ?? $this->defaultContext[self::FORCE_COLLECTION], true)
+        ) {
+            return [$childNodeName => [$value[$childNodeName]]];
+        }
```
This condition can go after the `!\is_array($value)` if.
Then, you'll be able to drop the `is_array` check.

AlexandreGerault replied:
Actually it can't according to the test suite 👀 
Looks like it actually passes the block below:
```php
if (!\count($data)) {
    return $value;
}
```

mtarld commented on src/Symfony/Component/Serializer/Encoder/XmlEncoder.php:
```diff
@@ -224,10 +230,20 @@ final protected function isElementNameValid(string $name): bool
+        if (\is_array($value)
+            && ($childNodeName = $node->firstChild?->nodeName)
+            && 1 === \count($value)
+            && \in_array($nodeName, $context[self::FORCE_COLLECTION] ?? $this->defaultContext[self::FORCE_COLLECTION], true)
```
I think it could be great to accept the value `*` as an "every tag" value, WDYT?

AlexandreGerault replied:
Maybe, however I'm still unsure how this PR would be welcome. I'm okay to continue working on it, but I'd like to get a green light before investing too much time on it 😅 

But if it get a positive feedback (which doesn't seem to be the case for the moment) I can consider adding a test case for this, maybe using xpath also 🤔

Commit:
feat(serializer): add FORCE_COLLECTION option to XmlEncoderContextBuilder

Commit:
feat(serializer): add static analysis

Hanmac commented:
You are using only the encode/decode functions, right?

Because for my project when i got something like
```xml
<order_rows>
    <order_row></order_row>
</order_rows>
```

I use `#[SerializedPath("[order_rows][order_row]")]`

AlexandreGerault commented:
@Hanmac what if your XML ends up looking like this:
```xml
<order_rows>
</order_rows>
```
Does it handle it well? :thinking:

Hanmac commented:
> @Hanmac what if your XML ends up looking like this:
> 
> ```
> <order_rows>
> </order_rows>
> ```
> 
> Does it handle it well? 🤔

That should be skipped

the only problem i have with this approach, (or in general) are empty collections when normalizing/encoding:

```php
  #[SerializedPath("[order_rows][order_row]")]
  protected array $rows = [];
```

if for some reason, the rows are empty, than current AbstractObjectNormalizer does normalize it to this:
```
[
'order_rows' => [
  'order_row' => []
]
]
```

which then the Encoder does turn into this:

```xml
<order_rows>
    <order_row />
</order_rows>
```

for this to work better, i added a Normalizer that turns empty arrays into `null` and then i have `AbstractObjectNormalizer::SKIP_NULL_VALUES` enabled to remove them.

i think i mentioned them in another ticket before, might need to create a new one.

AlexandreGerault commented:
Hey
I waited a bit and forgot about this PR. I notice the milestones have been updated since. Any feedback or interest in going further (working on how to define related tags for example, as suggested by @mtarld )?
I'm still up to work on this 👍

mtarld commented:
Hey @AlexandreGerault, I think you can go without bothering for the `*` for the moment :slightly_smiling_face: 
Let think about it in a new PR!

Commit:
[Serializer] Add a force_collection option to XmlEncoder

A tag holding a single child decodes to that child rather than to a list, so
the shape of the decoded data depends on how many children the document
happens to carry. Listing a tag under the force_collection context option
makes it always decode as a collection.
